spec/unit/puppetfile_spec.rb

```diff
@@ -68,34 +68,70 @@
-      subject = described_class.new(path, {})
+    context 'using load' do
+      it "returns the loaded content" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
 
-      loaded_content = subject.load
-      expect(loaded_content).to be_an_instance_of(Hash)
+        loaded_content = subject.load
+        expect(loaded_content).to be_an_instance_of(Hash)
 
-      has_some_data = loaded_content.values.none?(&:empty?)
-      expect(has_some_data).to be true
-    end
+        has_some_data = loaded_content.values.none?(&:empty?)
+        expect(has_some_data).to be true
+      end
 
-    it "is idempotent" do
-      path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
-      subject = described_class.new(path, {})
+      it "is idempotent" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
 
-      expect(subject.loader).to receive(:load).and_call_original.once
+        expect(subject.loader).to receive(:load).and_call_original.once
 
-      loaded_content1 = subject.load
-      expect(subject.loaded?).to be true
-      loaded_content2 = subject.load
+        loaded_content1 = subject.load
+        expect(subject.loaded?).to be true
+        loaded_content2 = subject.load
 
-      expect(loaded_content2).to eq(loaded_content1)
+        expect(loaded_content2).to eq(loaded_content1)
+      end
+
+      it "returns nil if Puppetfile doesn't exist" do
+        path = '/rando/path/that/wont/exist'
+        subject = described_class.new(path, {})
+        expect(subject.load).to eq nil
+      end
     end
 
-    it "returns false if Puppetfile doesn't exist" do
-      path = '/rando/path/that/wont/exist'
-      subject = described_class.new(path, {})
-      expect(subject.load).to eq false
+    context 'using load!' do
+      it "returns the loaded content" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
+
+        loaded_content = subject.load!
+        expect(loaded_content).to be_an_instance_of(Hash)
+
+        has_some_data = loaded_content.values.none?(&:empty?)
+        expect(has_some_data).to be true
+      end
+
+      it "is not idempotent" do
+        path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'valid-forge-with-version')
+        subject = described_class.new(path, {})
+
+        expect(subject.loader).to receive(:load).and_call_original.twice
+
+        loaded_content1 = subject.load!
+        expect(subject.loaded?).to be true
+        expect {
+          subject.load!
+        }.to raise_error(R10K::Error)
+      end
+
+      it "raises if Puppetfile doesn't exist" do
```

Comment:
Should this test go under `load` instead of `load!`?

Reply:
No, `load` will not raise, `load!` will.

Reply:
oh.. okay so yeah `load!` will raise, but we shouldn't ever actually hit this case, right? If everything calls `load` rather than calling `load!` directly, we'll hit the case where we log that the Puppetfile is missing or unreadable and never call `load!`. Am I missing something?

Reply:
No, the puppetfile actions, eg `R10K::Action::Puppetfile::Purge` all call `load!` and it is the ENOENT that was being raised by `puppetfile_content` that cause the action to abort, print a message at the ERROR level, and return 1. Nothing calls `load` within R10K itself, I don't really know why it exists.

Reply:
😑 fun.
Should we update those things to call `load` instead and then have `load` fail with a "nice" error that matches the debug message?

Reply:
No.

I think the refactor was a slight regression in UX and the implementation of that older, better UX was a little wonky. I've returned it to the better UX with a wonkier implementation. My end goal with this work it stop callers from using the Puppetfile and instead use the ModuleLoader impelementation (which behaves like `load!`) and have the callers check their inputs and handle them nicely.

I'd rather put a button on this part of the refactor and move on towards the eventual goal.

Reply:
okay, sounds good